src/map.rs

```diff
@@ -44,21 +44,65 @@ impl<K: Eq + Hash, V: StableDeref> FrozenMap<K, V> {
         ret
     }
 
-    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V::Target>
+    /// Applies a function to the owner of the value corresponding to the key (if any).
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use elsa::FrozenMap;
+    ///
+    /// let map = FrozenMap::new();
+    /// map.insert(1, Box::new("a"));
+    /// assert_eq!(map.map_get(&1, Clone::clone), Some(Box::new("a")));
+    /// assert_eq!(map.map_get(&2, Clone::clone), None);
+    /// ```
+    pub fn map_get<Q:?Sized, T, F>(&self, k:&Q, f:F) -> Option<T> 
     where
         K: Borrow<Q>,
         Q: Hash + Eq,
+        F: FnOnce(&V) -> T,
     {
         assert!(!self.in_use.get());
         self.in_use.set(true);
         let ret = unsafe {
             let map = self.map.get();
-            (*map).get(k).map(|x| &**x)
+            (*map).get(k).map(f)
         };
         self.in_use.set(false);
         ret
     }
 
+    /// Returns a reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use elsa::FrozenMap;
+    ///
+    /// let map = FrozenMap::new();
+    /// map.insert(1, Box::new("a"));
+    /// assert_eq!(map.get(&1), Some(&"a"));
+    /// assert_eq!(map.get(&2), None);
+    /// ```
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V::Target>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.map_get(k, |x| unsafe {
```

Comment:
I would keep the original implementation of this since we need the unsafe anyway